poly/promise/promise.go

```diff
@@ -0,0 +1,744 @@
+ *
+ * The BlameProof object provides an accountability measure. If a promiser
+ * decides to construct a faulty share, guardians can construct a BlameProof
+ * to show that the server is malicious. 
+ * 
+ * The guardian provides the index of the bad secret, the bad secret itself,
+ * and its diffie-hellman shared key with the promiser. Other servers can then
+ * verify if the promiser is malicious or the guardian is falsely accusing the
+ * server.
+ *
+ * To quickly summarize the blame procedure, two things must hold for the blame
+ * to succeed:
+ *
+ *   1. The provided share when encrypted with the diffie key must equal the
+ *   point provided at index bi of the promise. This verifies that the share
+ *   was actually intended for the guardian.
+ *
+ *   2. The provided share must fail to pass pubPoly.Check. This ensures that
+ *   the share is actually corrupted and the guardian is not just lying.
+ */
+
+type PromiseShare struct {
+
+	// The index of the share
+	i int
+	
+	// The actual share from the guardian
+	share abstract.Secret
+	
+	// The Diffie-Hellman key between guardian i and the promiser.
+	diffieKey abstract.Point
+}
+
+
+/* Initializes a new PromiseShare
+ *
+ * Arguments
+ *    i   = the index of the Promise share the guardian is revealing
+ *    s   = the share being revealed
+ *    d   = the Diffie-Hellman key between the guardian and promiser
+ *
+ * Returns
+ *   An initialized PromiseShare
+ */
+func (p *PromiseShare) Init(i int, s abstract.Secret, d abstract.Point) *PromiseShare {
+	p.i         = i
+	p.share     = s
+	p.diffieKey = d
+	return p
+}
+
+// Tests whether two promise secrets are equal.
+func (p *PromiseShare) Equal(p2 *PromiseShare) bool {
+	return p.i == p2.i && p.share.Equal(p2.share) &&
+	       p.diffieKey.Equal(p2.diffieKey)
+}
+
+/* Promise objects are mechanism by which servers can promise that a certain
+ * private key or abstract.Secret can be recomputed by other servers in case
+ * the original server goes offline.
+ *
+ * The Promise struct handles the logic of creating private shares, splitting
+ * these shares up for a given number of servers to act as guardians, verifying
+ * promises, and providing proof that guardians have indeed taken out approved
+ * of backing up the promse.
+ *
+ * Terms:
+ *   promiser = the server making the promise. The own who owns this object.
+ *   promised = recipients of the promise (aka the clients of the promiser)
+ *   guardian = another server who receives a share of the promise and can help
+ *              reconstruct it.
+ *
+ * Development note: The guardians, secrets, and signatures arrays should
+ *                   remain synchronized. In other words, the guardians[i],
+ *                   secrets[i], and signatures[i] should all refer to the same
+ *                   server.
+ */
+type Promise struct {
+
+	// The id of the promise. In the format:
+	//   PromiserPublicKey.String() + TimeOfCreation + RandomNumber
+	id string
```

Comment:
Initially, I was thinking this could help support multiple promises per server. The promise id could be used to differentiate them and as a hash key to allow quick look-ups. However, there are other ways of differentiating promises (namely the new Equal method I created). Furthermore, fast look-ups probably won't be much of a concern as a given server is unlikely to have too many active Promises at once.